src/slurm_plugin/computemgtd.py

```diff
@@ -0,0 +1,138 @@
+        # Log configuration
+        log.info(self.__repr__())
+
+
+def _self_terminate(computemgtd_config):
+    """Self terminate the instance."""
+    instance_manager = InstanceManager(
+        computemgtd_config.region, computemgtd_config.cluster_name, computemgtd_config.boto3_config
+    )
+    self_instance_id = check_command_output("curl http://169.254.169.254/latest/meta-data/instance-id")
+    log.info("Self terminating instance %s now!", self_instance_id)
+    instance_manager.delete_instances([self_instance_id])
+
+
+def _get_clustermgtd_heartbeat(clustermgtd_heartbeat_file_path):
+    """Get clustermgtd's last heartbeat."""
+    with open(clustermgtd_heartbeat_file_path, "r") as timestamp_file:
+        # Note: heartbeat must be written with datetime.strftime to convert localized datetime into str
+        # datetime.strptime will not work with str(datetime)
+        # Example timestamp written to heartbeat file: 2020-07-30 19:34:02.613338+00:00
+        return datetime.strptime(timestamp_file.read(), TIMESTAMP_FORMAT)
+
+
+def _run_computemgtd():
+    """Run computemgtd actions."""
+    # Initial default heartbeat time as computemgtd startup time
+    last_heartbeat = datetime.now(tz=timezone.utc)
+    log.info("Initializing clustermgtd heartbeat to be computemgtd startup time: %s", last_heartbeat)
+    while True:
+        # Get current time
+        current_time = datetime.now(tz=timezone.utc)
+        # Get program config
+        computemgtd_config = ComputemgtdConfig(os.path.join(CONFIG_FILE_DIR, "parallelcluster_computemgtd.conf"))
+        # Configure root logger
+        try:
+            fileConfig(computemgtd_config.logging_config, disable_existing_loggers=False)
+        except Exception as e:
+            log.warning(
+                "Unable to configure logging from %s, using default logging settings.\nException: %s",
+                computemgtd_config.logging_config,
+                e,
+            )
+        # Check heartbeat
+        try:
+            last_heartbeat = _get_clustermgtd_heartbeat(computemgtd_config.clustermgtd_heartbeat_file_path)
+            log.info("Latest heartbeat from clustermgtd: %s", last_heartbeat)
+        except Exception as e:
+            log.error("Unable to retrieve clustermgtd heartbeat with exception: %s", e)
+        finally:
+            # Any non-positive clustermgtd_timeout will disable self-termination
+            if computemgtd_config.clustermgtd_timeout > 0 and time_is_up(
+                last_heartbeat, current_time, computemgtd_config.clustermgtd_timeout
+            ):
+                log.error(
+                    "Clustermgtd has been offline since %s. Current time is %s. Timeout of %s seconds has expired!",
+                    last_heartbeat,
+                    current_time,
+                    computemgtd_config.clustermgtd_timeout,
+                )
+                _self_terminate(computemgtd_config)
```

Comment:
If I'm not wrong here we decided to have 2 timeouts/checks:
1. we check for the hearthbeat coming from clustermgtd
2. when the hearthbeat check fails I'd say we check for the status of the node in the scheduler. This because we do not want to terminate a node still connected to the scheduler even if the clustermgtd is temporarily offline. When checking the status of the scheduler we could issue a `scontrol show node nodename` where the nodename could be read from the computemgtd config and written in the chef recipes.

Also probably we want 2 timeouts:
1. on the hearth-beat not received in the last 5 mins (3 I think it's a bit too little in case clustermgtd is very busy)
2. on slurm failing to respond (we might want to retry here for a minute or so but with a very low rate, e.g. every 20 secs for 3 times)

Reply:
we need to handle failures in self_terminate by probably logging the error and having it retry at the next iteration rather than propagating the exception to the main program

Reply:
I will argue against the second slurm timeout, `SlurmdTimeout` should already take that into account, and I don't think we want to add a second timeout. Plus, timing on scheduler state is not straightforward, we will need another "pre-termination" loop to constantly check scheduler state(like nodewatcher), which is not that nice.

I think we should check the scheduler state only once whenever we see that the heartbeat is expired, and if the state is `down` or if we cannot retrieve scheduler state we will terminate.

Reply:
yes sorry I should have been more specific. The second timeout was referred to the fact that Slurm might be temporary unresponsive so we want a big enough timeout or at least retry for some time if the scheduler command fails (for example with a timeout). we cannot be super aggressive and terminate right away if get_nodes_info fails because of a timeout. We should give some time to recover as we do today in nodewatcher. If you recall we had to increase the timeout in nodewatcher to account for transient issues that make the scheduler unresponsive (e.g. network overload or scheduler overload)